src/aiida_quantumespresso/parsers/base.py

```diff
@@ -3,15 +3,78 @@
 
 All `Parser` implementations in `aiida-quantumespresso` must use this base class, not `aiida.parsers.Parser`.
 """
-from aiida.parsers import Parser as _BaseParser
+import abc
+import re
+from typing import Optional, Tuple
 
-__all__ = ('Parser',)
+from aiida.common import AttributeDict
+from aiida.engine import ExitCode
+from aiida.parsers import Parser
 
+from aiida_quantumespresso.parsers.parse_raw.base import convert_qe_time_to_sec
 
-class Parser(_BaseParser):  # pylint: disable=abstract-method
-    """Custom `Parser` class for `aiida-quantumespresso` parser implementations."""
+__all__ = ('BaseParser',)
 
-    def emit_logs(self, logging_dictionaries, ignore=None):
+
+class BaseParser(Parser, metaclass=abc.ABCMeta):
+    """Custom ``Parser`` class for ``aiida-quantumespresso`` parser implementations."""
+
+    class_error_map = {}
+    class_warning_map = {}
+
+    base_error_map = {
+        'Maximum CPU time exceeded': 'ERROR_OUT_OF_WALLTIME',
+    }
+    base_warning_map = {
+        'Warning:': None,
+        'DEPRECATED:': None,
+    }
+    success_string = 'JOB DONE'
+
+    @classmethod
+    def get_error_map(cls):
+        """The full error map of the parser class."""
+        error_map = cls.base_error_map.copy()
+        error_map.update(cls.class_error_map)
+        return error_map
+
+    @classmethod
+    def get_warning_map(cls):
+        """The full warning map of the parser class."""
+        warning_map = cls.base_warning_map.copy()
+        warning_map.update(cls.class_warning_map)
+        return warning_map
+
+    def parse_stdout_from_retrieved(self, logs: AttributeDict) -> Tuple[str, dict, AttributeDict]:
+        """Read and parse the ``stdout`` content of a Quantum ESPRESSO calculation.
+
+        :param logs: Logging container that will be updated during parsing.
+        :returns: size 3 tuple: (``stdout`` content, parsed data, updated logs).
```

Comment:
Why return the logs if they are passed in by reference and so are updated in place? Typically not a huge fan of tuple return values as it is often not immediately clear what is being returned.

Reply:
I do see your point. To me, I like making it clear that the `logs` input is updated by also passing it as an output. I guess it also related to the question I pose [here](https://github.com/aiidateam/aiida-quantumespresso/pull/756#discussion_r1192080727), and makes me inclined to simply generate separate logging containers in each parser method, but that also feels than ideal...

Maybe we should keep track of the logs on a class variable? That seems sensible. Initially I also wasn't sure why we needed this extra logging item because we can just actually use the `logger`. Then I figured this is so we can keep track of errors and then pick and choose which one to return as an exit code. But can't we just inspect the logger as well? Feels like we doing double work.

Re returning tuples: I can see your point, but what is the alternative for returning multiple outputs? A dictionary is even more opaque to me. Also: with the hover features in VSCode it's quite clear what is being returned: 🙃 

![Screenshot 2023-05-18 at 20 00 20](https://github.com/aiidateam/aiida-quantumespresso/assets/25607960/fea9bdd4-1124-4839-a721-b016fa2d6d27)

Reply:
You might not believe it, but I am actually VSCode myself now. But I think that in general making code design decisions based on a feature of the particular editor that we ourselves just happen to be using, is a bad idea, because not everyone is using the same tools. But fine, let's keep this as is.

Reply:
![Uploading good-palpatine.gif…]()